pkg/loadbalancerservice/controller.go

```diff
@@ -0,0 +1,154 @@
+	c.stopCh = make(chan struct{})
+	defer close(c.stopCh)
+
+	config, err := c.restConfig()
+	if err != nil {
+		return errors.Wrap(err, "error creating rest config for service controller")
+	}
+	c.client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create clientset for service controller")
+	}
+	c.ctx = ctx
+	close(ready)
+	go c.startServiceController(c.stopCh)
+
+	<-ctx.Done()
+
+	return ctx.Err()
+}
+
+func (c *LoadbalancerServiceController) startServiceController(stopped chan struct{}) error {
+	klog.Infof("Starting service controller")
+
+	factory := informers.NewSharedInformerFactory(c.client, defaultInformerResyncPeriod)
+	serviceInformer := factory.Core().V1().Services()
+	informer := serviceInformer.Informer()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAddService,
+		UpdateFunc: c.onUpdateService,
+	})
+
+	factory.Start(stopped)
+
+	if !cache.WaitForCacheSync(stopped, informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
+}
+
+func (c *LoadbalancerServiceController) restConfig() (*rest.Config, error) {
+	return clientcmd.BuildConfigFromFlags("", c.KubeConfig)
+}
+
+func (c *LoadbalancerServiceController) onAddService(obj interface{}) {
+	c.onUpdateService(nil, obj)
+}
+
+func (c *LoadbalancerServiceController) onUpdateService(oldObj, newObj interface{}) {
+	new := newObj.(*corev1.Service)
+	klog.Infof("process service %s/%s", new.Namespace, new.Name)
+	if new.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return
+	}
+	newStatus, err := c.getNewStatus(new)
+	if err != nil || newStatus == nil {
+		klog.Errorf("failed to get new service status: %s", err)
+		return
+	}
+	c.patchStatus(new, newStatus)
+}
+
+func (c *LoadbalancerServiceController) getNewStatus(svc *corev1.Service) (*corev1.LoadBalancerStatus, error) {
+	lbServices, err := c.client.CoreV1().Services(corev1.NamespaceAll).List(c.ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range lbServices.Items {
+		if (s.Namespace == svc.Namespace && s.Name == svc.Namespace) || s.Spec.Type != corev1.ServiceTypeLoadBalancer || len(s.Status.LoadBalancer.Ingress) == 0 {
+			continue
+		}
+		for _, ep := range s.Spec.Ports {
+			for _, np := range svc.Spec.Ports {
+				if ep.Port == np.Port {
+					return nil, fmt.Errorf("node port %d occupied", ep.Port)
+				}
+			}
+		}
+	}
+
+	newStatus := &corev1.LoadBalancerStatus{}
+	newStatus.Ingress = append(newStatus.Ingress, corev1.LoadBalancerIngress{
+		IP:       c.NodeIP,
+		Hostname: c.NodeName,
+	})
+	return newStatus, nil
+}
+
+func (c *LoadbalancerServiceController) patchStatus(svc *corev1.Service, newStatus *corev1.LoadBalancerStatus) error {
+	if helpers.LoadBalancerStatusEqual(&svc.Status.LoadBalancer, newStatus) {
+		return nil
+	}
+
+	updated := svc.DeepCopy()
+	updated.Status.LoadBalancer = *newStatus
+	_, err := helpers.PatchService(c.client.CoreV1(), svc, updated)
```

Comment:
This err is not logged anywhere, shall it be logged here or in the parent function (c.patchStatus) ?

Reply:
I've added an error-handling mechanism. The error will be logged in the function `handleErr`.